### PR TITLE
[timeseries] Enable Ray Tune for timeseries

### PR DESCRIPTION
**Pull request by canerturkmen**
*Issue #, if available:*

*Description of changes:*

We enable Ray Tune as the default HPO backend for GluonTS models except TFT. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 1]
```diff
@@ -384,6 +386,38 @@ def _get_hpo_train_fn_kwargs(self, **train_fn_kwargs) -> dict:
...
+    def _is_gpu_available(self) -> bool:
+        return False
+
+    def hyperparameter_tune(
+        self, hyperparameter_tune_kwargs="auto", hpo_executor: HpoExecutor = None, time_limit: float = None, **kwargs
+    ):
+        if hpo_executor is None:
+            hpo_executor = self._get_default_hpo_executor()
+            default_num_trials = kwargs.pop("default_num_trials", None)
+        hpo_executor.initialize(
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:398]: In `core.models.abstract.AbstractModel.hyperparameter_tune` this line is inside the `if` condition, why do we need to change it here?

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py [hunk 2]
```diff
@@ -429,21 +463,27 @@ def _hyperparameter_tune(
...
-            adapter_type="timeseries",
-        )
 
-        return hpo_executor.get_hpo_results(
-            model_name=self.name,
-            model_path_root=self.path_root,
-            time_start=time_start,
-        )
+        minimum_resources = self.get_minimum_resources(is_gpu_available=self._is_gpu_available())
+        hpo_context = disable_stdout if isinstance(hpo_executor, RayHpoExecutor) else nullcontext
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:468]: Nice!

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py [hunk 3]
```diff
@@ -302,6 +304,18 @@ def _get_estimator(self) -> GluonTSEstimator:
             **init_args,
         )
 
+    def _is_gpu_available(self) -> bool:
+        import torch.cuda
+
+        return torch.cuda.is_available()
+
+    def get_minimum_resources(self, is_gpu_available: bool = False) -> Dict[str, Union[int, float]]:
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:312]: Is my understanding correct?
- `get_minimum_resources` method is necessary to tell Ray "please always assign 1 GPU to this job if there is GPU available"
- all GluonTS models will always try to use the GPU if GPU is available, regardless of `num_gpus` passed to `model.fit` or anything else

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py [hunk 4]
```diff
@@ -58,6 +58,23 @@ def __init__(
...
+    def _get_model_base(self):
+        if self.most_recent_model is not None:
+            return self.most_recent_model
+        elif self.model_base is not None:
+            return self.model_base
+        else:
+            raise ValueError("Base model cannot be retrieved from MultiWindowModel")
+
+    def _get_hpo_backend(self) -> str:
+        return self._get_model_base()._get_hpo_backend()
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:70]: Why not just use `self.model_base._get_hpo_backend()`? The constructor of `MultiWindowBacktestingModel` should ensure that `self.model_base` is an object of type `AbstractTimeSeriesModel`. If  `self.model_base is None`, we will have even bigger problems (`get_child_model()` below will fail, so fit won't work).

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 5]
```diff
@@ -512,28 +512,36 @@ def fit(
...
-
-            Setting this parameter to string ``"random"`` performs 10 trials of random search per model.
-
-            We can change the number of random search trials per model by passing a dictionary as `hyperparameter_tune_kwargs`.
-            The dict must include the following keys
-
-            - ``"num_trials"``: int, number of configurations to train for each tuned model
-            - ``"searcher"``: currently, the only supported option is ``"random"`` (random search)
-            - ``"scheduler"``: currently, the only supported option is ``"local"`` (all models trained on the same machine)
+        hyperparameter_tune_kwargs : str or dict, default = None
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/predictor.py:515]: Nit: We usually use `str or dict, optional` to annotate the type `Optional[Union[str, dict]]`

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 6]
```diff
@@ -512,28 +512,36 @@ def fit(
...
+            Hyperparameter tuning strategy and kwargs (for example, how many HPO trials to run).
+            If None, then hyperparameter tuning will not be performed.
+
+            If type is ``str``, then this argument specifies a preset.
+                Valid preset values:
+                    'auto': Performs HPO via bayesian optimization search on GluonTS-backed neural forecasting models and
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/predictor.py:521]: Nit: The majority of the docstrings in this file user `"` instead of `'` to denote strings

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 7]
```diff
@@ -512,28 +512,36 @@ def fit(
...
-            The dict must include the following keys
-
-            - ``"num_trials"``: int, number of configurations to train for each tuned model
-            - ``"searcher"``: currently, the only supported option is ``"random"`` (random search)
-            - ``"scheduler"``: currently, the only supported option is ``"local"`` (all models trained on the same machine)
+        hyperparameter_tune_kwargs : str or dict, default = None
+            Hyperparameter tuning strategy and kwargs (for example, how many HPO trials to run).
+            If None, then hyperparameter tuning will not be performed.
+
+            If type is ``str``, then this argument specifies a preset.
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/predictor.py:519]: Will the nested paragraph structure be preserved once the docs are rendered? We might need to use bulleted lists to ensure that the text is displayed in a readable manner in the docs.

File: timeseries/src/autogluon/timeseries/predictor.py [hunk 8]
```diff
@@ -512,28 +512,36 @@ def fit(
...
+                Valid preset values:
+                    'auto': Performs HPO via bayesian optimization search on GluonTS-backed neural forecasting models and
+                        random search on other models using local scheduler.
+                    'random': Performs HPO via random search.
+            You can also provide a dict to specify searchers and schedulers
+                Valid keys:
+                    'num_trials': How many HPO trials to run
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/predictor.py:526]: Same as above, maybe use bulleted list for nicer rendering?

File: timeseries/tests/unittests/models/test_models.py [hunk 9]
```diff
@@ -412,6 +413,12 @@ def test_when_custom_metric_passed_to_model_then_model_can_hyperparameter_tune(m
         },
         eval_metric=CustomMetric(),
     )
+    backend = model._get_hpo_backend()
+    if backend is RAY_BACKEND:
+        # Ray has trouble keeping references to the custom metric in the test namespace. We therefore
+        # skip this test.
+        pytest.skip()
```
Comment (shchur) [inline timeseries/tests/unittests/models/test_models.py:420]: Can we do something like [this](https://auto.gluon.ai/stable/tutorials/tabular/advanced/tabular-custom-metric.html#ensuring-metric-is-serializable) to ensure that Ray works with our implementation of custom metrics?

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py (hunk 1 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:398]: Thanks for catching this!

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py (hunk 4 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:70]: `._get_model_base()` is an `AbstractModel` method that ray also relies on (assuming that when the base model is not equal to the main model, then you're using bagging) and `AbstractModel._get_default_hpo_executor()` expects to be working properly. I can have `_get_model_base()` just return `self.model_base` but I would rather keep the idiom of `_get_model_base` for consistency. wdyt?

File: timeseries/src/autogluon/timeseries/predictor.py (hunk 5 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/predictor.py:515]: sure. but, how do we imply it's None by default?

File: timeseries/src/autogluon/timeseries/predictor.py (hunk 7 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/predictor.py:519]: This mirrored Tabular. I updated this and will check the rendering.

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py (hunk 3 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:312]: That matches my understanding, yes.

File: timeseries/src/autogluon/timeseries/predictor.py (hunk 5 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/predictor.py:515]: If the default value was not None, we would write 
```
hyperparameter_tune_kwargs : str or dict, default = "my_default_value"
```
while 
```
hyperparameter_tune_kwargs : str or dict, optional
```
implies that the default value is None. At least that's how it's used in the rest of TimeSeriesPredictor / TimeSeriesDataFrame documentation.

This is also a convention that Numpy uses - see, e.g., docs for `dtype` and `like` arguments [here](https://numpy.org/doc/stable/reference/generated/numpy.ones.html)

File: timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py (hunk 4 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/multi_window/multi_window_model.py:70]: Always returning `self.model_base` from `_get_model_base()` sounds good to me.

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py (hunk 3 above)
Comment (shchur) [reply timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:312]: Thanks!

File: timeseries/tests/unittests/models/test_models.py (hunk 9 above)
Comment (canerturkmen) [reply timeseries/tests/unittests/models/test_models.py:422]: Hmm :/ I don't think so. I don't think the problem is serializability, but rather the fact that during tests Ray doesn't include timeseries.tests.unittests in the pythonpath and find the `CustomMetric` object to unpickle with. If I'm thinking about it, I don't think there's anything more serializable / unserializable about `CustomMetric` than our own metric implementations. I haven't had any trouble so far on this end.

**Review by shchur — APPROVED** [review]
LGTM, thank you! Feel free to merge after the platform tests pass
